File: labs/Module_3_JAVA_LAB_2/Triple_Modular_lucky_array.java
import java.util.Arrays;
import java.util.Scanner;

class Triple_Modular_lucky_array {
    public static int tripleModularExponentiation(int[] arr) {
        int mod = 9;
        int result = 1;
        for (int num : arr) {
            result = (((result * num) % mod) * (int) Math.pow(result, mod)) % mod;
        }
        return result;
    }

    public static String findLuckyArray(int[] arr) {
        int n = arr.length;
        if (n < 4) {
            return "Invalid Input";
        }

        Arrays.sort(arr); // Sort the array in ascending order

        int maxModulus = 0;
        int[] luckyArray = new int[4];

        for (int i = 0; i <= n - 4; i++) {
            int[] subarray = Arrays.copyOfRange(arr, i, i + 4);
            int modulus = tripleModularExponentiation(subarray);
            if (modulus > maxModulus) {
                maxModulus = modulus;
                luckyArray = subarray;
            }
        }

        if (maxModulus == 0) {
            return "There is no lucky array in the list of subarrays";
        } else {
            StringBuilder sb = new StringBuilder();
            sb.append("[");
            for (int i = 0; i < luckyArray.length; i++) {
                sb.append(luckyArray[i]);
                if (i != luckyArray.length - 1) {
                    sb.append(", ");
                }
            }
            sb.append("] ");
            sb.append(maxModulus);
            return sb.toString();
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();
        String[] inputArray = input.split(" ");
        int[] arr = new int[inputArray.length];
        for (int i = 0; i < inputArray.length; i++) {
            arr[i] = Integer.parseInt(inputArray[i]);
        }

        System.out.println(findLuckyArray(arr));
    }
}
                                                                                                                            

File: labs/Module_3_Python_2/Triple_Modular_lucky_array.py

def find_lucky_array(arr, m=9):
    n = len(arr)
    if n < 4:
        print("Invalid Input")
        return

    arr.sort()  # Sort the array in ascending order

    max_modulo = -1  # Initialize max_modulo to -1
    lucky_arr = []
    multiple_lucky_arrays = False

    for i in range(n - 3):
        a, b, c, d = arr[i:i + 4]
        modulo = ((((a * b) % m) ** c) % m) ** c % m
        if modulo > max_modulo:
            max_modulo = modulo
            lucky_arr = [a, b, c, d]
            multiple_lucky_arrays = False
        elif modulo == max_modulo:
            multiple_lucky_arrays = True

    if multiple_lucky_arrays:
        print("There is more than one lucky arrays")
    elif lucky_arr:
        print(*lucky_arr, max_modulo)
    else:
        print("There is no lucky array in the list of subarrays")

# Example usage
find_lucky_array([2, 4, 5, 3, 1]) 